lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1144 @@
+///   `Reader` and `Writer` both hold a `SyncSocketDescriptor` clone.
+fn init_channels() -> (
+    Sender<ReaderCommand>,
+    Receiver<ReaderCommand>,
+    Sender<WriterCommand>,
+    Receiver<WriterCommand>,
+    Sender<Vec<u8>>,
+    Receiver<Vec<u8>>,
+) {
+    let (reader_cmd_tx, reader_cmd_rx) = crossbeam_channel::unbounded();
+    let (writer_cmd_tx, writer_cmd_rx) = crossbeam_channel::unbounded();
+    let (write_data_tx, write_data_rx) = crossbeam_channel::bounded(1);
+
+    (
+        reader_cmd_tx,
+        reader_cmd_rx,
+        writer_cmd_tx,
+        writer_cmd_rx,
+        write_data_tx,
+        write_data_rx,
+    )
+}
+
+/// A concrete implementation of the SocketDescriptor trait.
+///
+/// A SyncSocketDescriptor is essentially a `clone()`able handle to an
+/// underlying connection as well as an identifier for that connection.
+///
+/// This type is public only because handle_connection() requires it to be.
+#[derive(Clone)]
+pub struct SyncSocketDescriptor {
+    id: u64,
+    tcp_disconnector: TcpDisconnectooor,
+    reader_cmd_tx: Sender<ReaderCommand>,
+    writer_cmd_tx: Sender<WriterCommand>,
+    write_data_tx: Sender<Vec<u8>>,
+}
+impl PartialEq for SyncSocketDescriptor {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+impl Eq for SyncSocketDescriptor {}
+impl hash::Hash for SyncSocketDescriptor {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.id.hash(state)
+    }
+}
+impl SyncSocketDescriptor {
+    fn new(
+        connection_id: u64,
+        tcp_disconnector: TcpDisconnectooor,
+        reader_cmd_tx: Sender<ReaderCommand>,
+        writer_cmd_tx: Sender<WriterCommand>,
+        write_data_tx: Sender<Vec<u8>>,
+    ) -> Self {
+        Self {
+            id: connection_id,
+            tcp_disconnector,
+            reader_cmd_tx,
+            writer_cmd_tx,
+            write_data_tx,
+        }
+    }
+}
+impl SocketDescriptor for SyncSocketDescriptor {
+    /// Attempts to queue up some data from the given slice for the `Writer` to
+    /// send. Returns the number of bytes that were successfully pushed to the
+    /// `write_data` channel.
+    ///
+    /// This implementation never calls back into the PeerManager directly,
+    /// thereby preventing reentrancy / deadlock issues. Instead, any commands
+    /// to be processed and data to be sent are dispatched to the Reader or
+    /// Writer via crossbeam channels.
+    ///
+    /// Additionally, sending across the crossbeam channels is done exclusively
+    /// with non-blocking try_send()s rather than blocking send()s, to ensure
+    /// that this function always returns immediately, thereby also reducing the
+    /// amount of time that the PeerManager's internal locks are held.
+    fn send_data(&mut self, data: &[u8], resume_read: bool) -> usize {
+        if resume_read {
+            let _ = self.reader_cmd_tx.try_send(ReaderCommand::ResumeRead);
```

Comment:
maybe pull these out into functions like

```rust
fn resume_read(&self) {
  let _ = self.reader_cmd_tx.try_send(ReaderCommand::ResumeRead);
}

fn pause_read(&self) {
  let _ = self.reader_cmd_tx.try_send(ReaderCommand::PauseRead);
}
```

Reply:
Those functions wouldn't be reused anywhere in the SyncSocketDescriptor so I have to disagree here